packages/core/src/util/array_utils.ts

```diff
@@ -24,22 +24,62 @@ export function addAllToArray(items: any[], arr: any[]) {
  * Flattens an array.
  */
 export function flatten(list: any[], dst?: any[]): any[] {
-  if (dst === undefined) dst = list;
-  for (let i = 0; i < list.length; i++) {
-    let item = list[i];
+  if (dst === undefined) dst = [];
+  flattenIntoExisting(list, dst);
+  return dst;
+}
+
+/**
+ * Flatten `src` array into `dst` array and computes if differences were detected.
+ *
+ * @param src Source array which can be hierarchical (array within array)
+ * @param dst Destination array which can have existing values.
+ * @returns `true` if `dst` array was muted in order to represent flatten version of `src`.
+ */
+export function flattenIntoExisting(src: any[], dst: any[]): boolean {
+  let dstIndx = flattenIntoExistingRecursive(src, dst, 0);
+  let differenceDetected = dstIndx < 0;
+  dstIndx = differenceDetected ? ~dstIndx : dstIndx;
+  while (dstIndx < dst.length) {
+    dst.pop();
+    differenceDetected = true;
+  }
```

Comment:
Could the following be faster?

```suggestion
  if (dstIndx < dst.length) {
    // Fewer items than were previously in `dst`, so trim the array.
    dst.length = dstIndex;
    differenceDetected = true;
  }
```

Reply:
fixed